### ISSUE TITLE
[HOLD for payment] [$2000] Wrong chat composer highlighted after clicking out of emoji modal

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14848]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open app, access a report
2. Press the edit message icon
3. Press the emoji icon in the edit composer to open the emoji modal
4. Exit the emoji model by pressing outside

## Expected Result:
The main composer is not highlighted, and the keyboard is focused on the editing composer

## Actual Result:
The main composer is highlighted, and the keyboard is not focused on the editing composer

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [x] iOS / Safari

**Version Number:** 1.2.65-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/216879422-75923acb-656c-4b27-8a56-d15e7ec68a2c.mp4

https://user-images.githubusercontent.com/43996225/216879478-34c57868-40a7-4ec3-be7e-88380da3f4b8.MP4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @Tienifr

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1675481157349539

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01aa19f44ae1aa2e68</li>
        <li>Upwork Job ID: 1623817661403475968</li>
        <li>Last Price Increase: 2023-02-16</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
@davidcardoza Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by davidcardoza** [comment]
I was able to reproduce the issue. Going to bring this out to a contributor.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @NikkiWines (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01aa19f44ae1aa2e68

**Comment by MelvinBot** [comment]
Current assignee @davidcardoza is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @mollfpr (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @cristipaval (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by cristipaval** [comment]
Why did Melvin assign both me and @NikkiWines ? 👀

**Comment by allroundexperts** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When the emoji selector popup is opened from the edit message composer and then closed, the main chat composer gets the focus instead of the edit message composer.

### What is the root cause of that problem?

The main composer has `autoFocus` property partially set to `true`.  That is, whenever the main composer mounts, it gets focused due to the presence of this property. 

On smaller screens, when the second composer (one which comes up when you edit the message) gets the focus, we're un-mounting the main composer. When the secondary composer looses focus (by a click outside), we then mount the main composer again. Since the main composer has auto-focus property setup, the main composer gets focused. This can be seen in the `ReportActionItemMessageEdit` file. The function used to show or hide the main composer is `toggleReportActionComposeView`.

When the emoji popup gets closed by clicking outside the secondary composer, it gets blur for a split second. At the same instance, the emoji popup also calls `onModalHide` since it has been closed. The two events happen almost at the same time. The first event (secondary composer blur) causes the primary composer to show up and be focused. The second event (onModalHide) causes the secondary composer to focus. The first event takes more time to complete (since it mounts a whole component) and thus, causes the main composer to be in a focused state.


### What changes do you think we should make in order to solve the problem?

We can do the following:
1. We can delay the callback executed in the `onModalHide` function in the emoji picker popup. This will make sure that the secondary composer focus happens AFTER the main composer has finished its focus. For an example, we can use `setTimeout` function for this.

### Alternate Solutions
- We could probably remove this functionality of hiding the main composer when the secondary composer gets focus on small screens altogether. It's more of a design / UX decision though.
- We can also remove `autofocus` property of the main composer. This will cause the `onModalHide` event to always execute which is what we want in this case.

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
After closing the emoji picker when editing message, the wrong composer was focused instead of the edit composer

### What is the root cause of that problem?
After the emoji picker is closed, the focus was returned to the `body` element, at the same time we try to execute `focus` on the edit composer. This causes a race condition, when the event `focus was returned to the body element` happens **after** `the edit composer is focused`, it causes the edit composer to immediately be blurred and triggered the main composer to appear and gain focus.

### What changes do you think we should make in order to solve the problem?

In `src/components/EmojiPicker/EmojiPicker.js`, we need to add a new function (only for web) for `onModalHide` that wraps the function passed in to the EmojiPicker. This new function will create a one-time listener for the `focusin` event of the `body` element in the DOM and only run the original `onModalHide` after the body is already focused, thus preventing the race condition.

Here's the pseudo code of the wrapped function:
```
wrappedOnModalHide(onModalHide) {
        return () => {
            if (document) {
                document.addEventListener('focusin', (event) => {
                    if (event.target.localName === 'body') {
                        onModalHide();
                    }
                    
                }, { once: true });
            }
        }
    }
```

We can optionally consider putting that same logic down to the `Modal` component instead if we want this to apply to all modals.

### What alternative solutions did you explore? (Optional)

We can use timeout but generally if there's a working solution that doesn't use timeout, we should prioritize that solution.


### Result
Working well after the fix:

https://user-images.githubusercontent.com/113963320/218247085-e8e424a5-d1d0-48b5-9573-b35caf53901c.mov

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When editing the personal message text input of `Invite new members` page, we can't use keyboard arrow-up and arrow-down to move the text input cursor. But it does change the `focusedIndex` of member selection option list.

### What is the root cause of that problem?

The root cause is because, in the capturing phase, 

1. the `document` captures arrow-up and arrow-down events triggered from message TextInput 
2. registered callback and `event.preventDefault()` get called in the handler [bindHandlerToKeydownEvent](https://github.com/Expensify/App/blob/d42e443d35762f15d1ebd3181bbc17636dff3d98/src/libs/KeyboardShortcut/index.js#L68)

which makes the message text input can't react to arrow-up or arrow-down event but change the focused item of the selection option list.

If we press arrow-up or arrow-down in the message text input, the `event.preventDefault()` will be called at the first phase of the 3 DOM event phases

```
Capturing phase => Target phase => Bubbling phase
```

So we can't perform default arrow-up or arrow-down action to move cursor in the text input.



In the codebase, we suscribe `keydown` events and get callbacks called in the capturing phase. Let's see how it works

1. We let the `document` listen `keydown` events through https://github.com/Expensify/App/blob/d42e443d35762f15d1ebd3181bbc17636dff3d98/src/libs/KeyboardShortcut/index.js#L107
2. In components, we can subscribe `keydown` events through `KeyboardShortcut.subscribe` 
3. If there's a `keydown` event, the handler [bindHandlerToKeydownEvent](https://github.com/Expensify/App/blob/d42e443d35762f15d1ebd3181bbc17636dff3d98/src/libs/KeyboardShortcut/index.js#L68) will be called in the capturing phase to decide how to call registered callbacks and `event.preventDefault()` based on event subscription configurations



Back to this issue, in the `Invite new members` page, we use `ArrowKeyFocusManager`  

```
WorkspaceInvitePage.js => OptionsSelector => BaseOptionsSelector.js => ArrowKeyFocusManager.js
```

 to subscribe arrow-up and arrow-down events with following two key properties set

```javascript
captureOnInputs = true
shouldPreventDefault = true // default value
```

See API doc of [KeyboardShortcut.subscribe](https://github.com/Expensify/App/blob/d42e443d35762f15d1ebd3181bbc17636dff3d98/src/libs/KeyboardShortcut/index.js#L138-L151)

So when pressing keyboard arrow-up or arrow-down in the message input, in the capturing phase, the handler [bindHandlerToKeydownEvent](https://github.com/Expensify/App/blob/d42e443d35762f15d1ebd3181bbc17636dff3d98/src/libs/KeyboardShortcut/index.js#L68) will call registered callback of arrow-up or arrow-down and also call `event.preventDefault()` because we use the default setting `shouldPreventDefault = true` when subscribing the event. See 

https://github.com/Expensify/App/blob/d42e443d35762f15d1ebd3181bbc17636dff3d98/src/libs/KeyboardShortcut/index.js#L93-L98

That's why we can't move the cursor in the message input but can change the focused item of the selection option list.

### What changes do you think we should make in order to solve the problem?

To fix this issue, we can 

1. Subscribe arrow-up and arrow-down events in `ArrowKeyFocusManager.js` with `shouldPreventDefault = false` to allow default behaviour to move cursor up and down in text input.

2. Disable interactivity of `WorkspaceInvitePage.js => OptionsSelector` when the personal message text input is focused. We can

   1. Add a state `this.state.isMessageTextInputFocused` in `WorkspaceInvitePage`

   2. Set  `disableArrowKeysActions` for `OptionsSelector` to skip the action of arrow-up and arrow-down shortcut

   3. Set `isDisabled` for `OptionsSelector` and check if `isDisabled` is set in the callback of `enter` shortcut in `BaseOptionsSelector`, like

      ```javascript
      this.unsubscribeEnter = KeyboardShortcut.subscribe(
          enterConfig.shortcutKey,
          () => {
              if (this.props.isDisabled) {
                  return;
              }
      ```

      to skip the action of `enter` shortcut

### What alternative solutions did you explore? (Optional)

NA

### Result

Working well after the fix

https://user-images.githubusercontent.com/117511920/218298792-e94b2e1c-3208-4f65-962b-3e13213487bd.mp4

**Comment by s77rt** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Closing the emoji modal does not focus the corresponding composer.

### What is the root cause of that problem?
The [onModalHide](https://github.com/Expensify/App/blob/f8d7a3a8e291cd216b26b22e9e8a69c93bb15e9e/src/pages/home/report/ReportActionItemMessageEdit.js#L260) callback is being called too early (i.e. before actually hiding the modal). This is a bug in `react-native-modal`. Focusing the composer before hiding the modal is illegal where [focus trap](https://github.com/Expensify/react-native-web/blob/master/packages/react-native-web/src/exports/Modal/ModalFocusTrap.js) kicks in and [shifts](https://github.com/Expensify/react-native-web/blob/e9b7517663f996e123772b0047a6db3415ed59da/packages/react-native-web/src/exports/Modal/ModalFocusTrap.js#L113-L126) the focus back to the modal.

### What changes do you think we should make in order to solve the problem?
- Fix the bug upstream: https://github.com/react-native-modal/react-native-modal/pull/733

### What alternative solutions did you explore? (Optional)
- Reach out to RN to implement a `onHide` prop and wait for RNW to adopt it, then use that new prop on `react-native-modal` to determine when to call the `onModalHide` callback.

**Comment by allroundexperts** [comment]
@s77rt Is this really a bug upstream? Aren't you just delaying the callback as well but in a different way? (Rendering twice vs using setTimeout)?

**Comment by s77rt** [comment]
@allroundexperts Thanks for taking the time to consider my proposal. I believe we can handle such discussion on slack. I have created this [thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1676216263906999). Feel free to post your concerns there.

**Comment by mollfpr** [comment]
Thanks guys! I’ll start reviewing the proposal.

**Comment by mollfpr** [comment]
@allroundexperts We should avoid using the `setTimeout` for a solution. Also, we already delay the input focus with `InteractionManager.runAfterInteractions`, which already works fine for other platforms.

@s77rt The upstream already has the [issue](https://github.com/react-native-modal/react-native-modal/issues/534) reported, but it's closed, and in the other issue, the maintainer [recommends](https://github.com/react-native-modal/react-native-modal/issues/484#issuecomment-812912374) to use `InteractionManager.runAfterInteractions`. Before opening a PR in E/App or the upstream library, we should discuss the solution first. Although your PR is solving the issue logically, we still need to determine if the maintainer will agree with the solution when it looks like a hack.

@tienifr Your result is promising, and the solution seems straightforward, but I still need to figure out if this is the best we can do.

@ntdiary Surprisingly, `onDismiss` works! Is there any explanation for why it works?

**Comment by allroundexperts** [comment]
@mollfpr We're making use of this at several places to solve this same problem. See here: https://github.dev/Expensify/App/blob/f40f9c9d81547292a53da4e48c23c2151a9d7fb7/src/pages/home/report/ReportActionCompose.js#L353

Also, `runAfterInteractions` is not a delay to be precise. The above mentioned function uses this inside of the `runAfterInteractions` as well.

**Comment by mollfpr** [comment]
@allroundexperts Yes, that will be the last option to consider. For now, it's too early to decide using `setTimeout`.

**Comment by s77rt** [comment]
@mollfpr Thanks for the review! I believe that using `runAfterInteractions` is just a [hack](https://expensify.slack.com/archives/C01GTK53T8Q/p1676238262426119) and the input should be focusable even without it.
@davidcardoza @cristipaval Can you help us get a response from the maintainers for this [PR](https://github.com/react-native-modal/react-native-modal/pull/733).

**Comment by tienifr** [comment]
@ntdiary just curious how you were able to make it work consistently with `onDismiss`, I tried but it seems to work some of the times and not work in some others.

**Comment by tienifr** [comment]
@s77rt I think if anything needs to be fixed upstream the fix should be on `react-native-web` rather than `react-native-modal`. 

`react-native-modal` doesn't know anything about `ModalFocusTrap` for example, so I'm not sure if `react-native-modal` should be updated due to an implementation detail of another library (`react-native-web`). I think `react-native-modal` works well on native Android, iOS where it's originally intended.

**Comment by s77rt** [comment]
@tienifr Thanks for your input. Would you please take that to Slack. Let's keep GH for proposals only and Slack for the discussion.

**Comment by mollfpr** [comment]
>@ntdiary just curious how you were able to make it work consistently with onDismiss? 😁
>
>I tried (by changing onModalHide to onDismiss in src/components/Modal/BaseModal.js) but it seems to work some of the times and not work in some others.

@tienifr This is interesting. Could you share the step you use? Is it happening on an iOS safari simulator or a physical device?

---

>@mollfpr, The children of ModalAnimation are removed immediately after onDismiss is executed (same loop).

@ntdiary I'm unsure what `ModalAnimation` need to do to make `onDismiss` work well than `onModalHide`. Could you elaborate?

@ntdiary I see you have a step on the focus trap issue in RNW https://github.com/necolas/react-native-web/issues/2483. Let's say that they agree to have a prop to control the focused element after the modal it's closed; I assume that the update we need to do is just add the element we want to focus on. Is it right?

**Comment by tienifr** [comment]
> I tried (by changing onModalHide to onDismiss in src/components/Modal/BaseModal.js) but it seems to work some of the times and not work in some others.

@mollfpr @ntdiary I'm no longer able to reproduce this issue, perhaps I missed out something, pls ignore it 🙏

**Comment by mollfpr** [comment]
@ntdiary Thanks for the clear explanation!

---

@cristipaval We can go with @ntdiary [proposal](https://github.com/Expensify/App/issues/14848#issuecomment-1428032631), but only replacing `onModalHide` with `onDismiss` for the web.


🎀 👀 🎀 C+ reviewed!

**Comment by cristipaval** [comment]
Thank you @mollfpr for triaging the proposals! I am happy with your choice, @ntdiary please go ahead with the fix 🚀

**Comment by s77rt** [comment]
@cristipaval Can you assign @ntdiary to this issue? (to remove the "Help Wanted" label)

**Comment by MelvinBot** [comment]
📣 @ntdiary You have been assigned to this job by @davidcardoza!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by davidcardoza** [comment]
Done.

**Comment by situchan** [comment]
[PR](https://github.com/Expensify/App/pull/15298) which fixes this issue caused regression - https://github.com/Expensify/App/issues/15559

**Comment by cristipaval** [comment]
> [PR](https://github.com/Expensify/App/pull/15298) which fixes this issue caused regression - #15559

Thanks @situchan !

We discussed a bit [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1677598279802909) on Slack. 
The workaround that I was thinking about is to set `shouldUseOnDismiss` by default as it is set in the @ntdiary's PR, but also make it possible to set another value for it in specific situations where we see at testing that it doesn't work with the default value. In our [regression](https://github.com/Expensify/App/issues/15559), if we override the default value and set `shouldUseOnDismiss` as `false`, it works as expected.
cc @s77rt

**Comment by MelvinBot** [comment]
⚠️ Looks like this issue was linked to a **Deploy Blocker** [here](https://github.com/Expensify/App/issues/15559)

If you are the assigned CME please investigate whether the linked PR caused a regression and leave a comment with the results.

If a regression has occurred and you are the assigned CM follow the instructions [here](https://stackoverflow.com/c/expensify/questions/13647).

If this regression could have been avoided please consider also proposing a recommendation to the PR checklist so that we can avoid it in the future.

**Comment by s77rt** [comment]
@ntdiary Sorry to see things didn't go as expected. If I may I'd like to re-propose my solution.
@mollfpr @cristipaval Given that the chosen solution is no longer generic, I'd appreciate a second look on this [one](https://github.com/Expensify/App/issues/14848#issuecomment-1427020356).

**Comment by s77rt** [comment]
Thanks for testing @mollfpr I'm not really sure why you are experiencing such delay, maybe it's unrelated to the change, can you test more?

The upstream bug is that the callback is being called too early, so it's natural that the fix will introduce some delay, but in your video the delay looks to be too much. If it may help try remove the `runAfterInteractions` call [here](https://github.com/Expensify/App/blob/f8d7a3a8e291cd216b26b22e9e8a69c93bb15e9e/src/pages/home/report/ReportActionItemMessageEdit.js#L260) and focus directly.  `runAfterInteractions` is not required here especially since it's a [hack](https://expensify.slack.com/archives/C01GTK53T8Q/p1676238262426119), but maybe we should keep it for native only.

**Comment by mollfpr** [comment]
@s77rt Yeah, I just removed `runAfterInteractions,` and there's no delay anymore. 

Your PR on RNM has been running for about 2 weeks, and has yet to receive a comment from the maintainer, and we are still determining if they will approve your solution. So it will take longer to solve this issue if we go with your solution.

There's also a chance that we will re-evaluate the solution if they do not approve your solution.

---

@tienifr We have yet to decide, but if we want to re-evaluate the proposal. I will review all the proposals again.

**Comment by mollfpr** [comment]
@cristipaval I'm still thinking the `onDismiss` is still a valid solution and @ntdiary [explain](https://github.com/Expensify/App/issues/14848#issuecomment-1449783536) have a solid point for what we have missing.

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.77-4 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15298
- https://github.com/Expensify/App/pull/15498


If no regressions arise, payment will be issued on 2023-03-09. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@mollfpr / @cristipaval] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@mollfpr / @cristipaval] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@mollfpr / @cristipaval] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@ntdiary] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@davidcardoza] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by mollfpr** [comment]
@ntdiary Yup, we will get the answer shortly!

**Comment by tienifr** [comment]
@cristipaval what do you think about my proposal [here](https://github.com/Expensify/App/issues/14848#issuecomment-1426652590)? It's a straight forward fix.

**Comment by allroundexperts** [comment]
@cristipaval Please comment on mine proposal [here](https://github.com/Expensify/App/issues/14848#issuecomment-1426512607) as well. We seem to be using this technique at several places.

**Comment by cristipaval** [comment]
Thank you for your proposals @tienifr @allroundexperts ! I saw your proposals but I think @mollfpr already did a great job by triaging all proposals and now I have the same feelings as him, weighing up the 2 solutions mentioned in my previous comment.

@ntdiary Thanks for jumping in with a new suggestion (looks like I have to delay the internal discussion a bit). @mollfpr what do you think about it?

**Comment by mollfpr** [comment]
@ntdiary Sorry for the delay, and thank you for the ping!

**Comment by mollfpr** [comment]
@ntdiary Your solution [here](https://github.com/Expensify/App/issues/14848#issuecomment-1460381959) is working fine solving this issue, but it's not solving the similar issue with opening the attachment modal.

https://user-images.githubusercontent.com/25520267/224722437-542a93d4-293a-4830-a264-e277815197a7.mp4

@cristipaval Should we also fix it for the modal attachment issue?

**Comment by mollfpr** [comment]
@ntdiary Ahh yeah, that's correct. Sorry about the unclear comment 😅

**Comment by cristipaval** [comment]
not overdue

**Comment by cristipaval** [comment]
Ooooh @ntdiary ! Now after watching carefully your screen recordings again, it seems to me that the bug related to the attachment modal is a different one. Even the reproducible steps are different. Given the fact that we have to make a ux decision in order to fix it, I think we better create a separate issue for it and tag the design team. Let's fix the original problem of this issue for now. Thank you for bearing with me. cc @mollfpr

**Comment by mollfpr** [comment]
@cristipaval I'm okay with @ntdiary new proposal https://github.com/Expensify/App/issues/14848#issuecomment-1460381959, but I'm not sure how you feel about re-render the component. Also, it's similar to what @s77rt proposed earlier to the upstream. The difference is we call it in our component, but the @ntdiary explanation makes more sense.

**Comment by cristipaval** [comment]
> @cristipaval I'm okay with @ntdiary new proposal https://github.com/Expensify/App/issues/14848#issuecomment-1460381959, but I'm not sure how you feel about re-render the component.

@mollfpr Do you have any concerns about re-rendering the modal just for safari on mobile? I know this isn't ideal but I wonder if there will be any lags or any visible bad experience for the user.

**Comment by MelvinBot** [comment]
@davidcardoza, @ntdiary, @cristipaval, @mollfpr Huh... This is 4 days overdue. Who can take care of this?

**Comment by cristipaval** [comment]
@mollfpr Please see my question ☝️

**Comment by mollfpr** [comment]
@cristipaval Sorry, I miss this issue. 

I think it’s okay to do it only on Mweb safari, but I think it’s better to push it to the Slack channel and see what the other thinks about the approach.

**Comment by cristipaval** [comment]
Slack discussion [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1680619388865949)

**Comment by cristipaval** [comment]
According to the Slack discussion [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1680619388865949), I think we should keep the `Help Wanted` label and wait for some more proposals on this one.

@ntdiary , it looks like you still have ideas to solve this issue, like [this one](https://expensify.slack.com/archives/C01GTK53T8Q/p1680703478276889?thread_ts=1680619388.865949&cid=C01GTK53T8Q). Could you please post them here using the [Proposal Template](https://github.com/Expensify/App/blob/main/contributingGuides/PROPOSAL_TEMPLATE.md)?

**Comment by mollfpr** [comment]
Thanks @ntdiary with the update, I'll review the updated proposal in the morning!

**Comment by mollfpr** [comment]
@cristipaval I'm unsure if replacing `InteractionManager.runAfterInteractions` with `setImmediate` is the right solution, even though it's fixed this issue.

**Comment by mollfpr** [comment]
>As for if we want to remove the setTimeout(xxx.focus(), 100) on android, it's probably beyond of the scope of this issue, but I can submit a PR to react-native to fix it if needed.

This will be great! What is your idea for the PR to RN upstream?

**Comment by mollfpr** [comment]
So implementing `onDismiss` for Android in `react-native` will make `react-native-modal` `onModalHide` work well? How about the `onModalHide` implementation in `react-native-web`?

**Comment by mollfpr** [comment]
>As for web platform, I think we will finally upgrade to react-native-web 0.19.x version and enable the concurrent mode,
at which point onModalHide or onDismiss will no longer be a big deal since they will be placed in the same macro task. We just need to fix the regression https://github.com/Expensify/App/issues/15559 with another way.

Let's focus on this.

> We just need to fix the regression https://github.com/Expensify/App/issues/15559 with another way.

@ntdiary do you mean that the regression is resurfaced again? Did we already revert the PR? I'm a bit lost here 😅

**Comment by mollfpr** [comment]
>@mollfpr, IMO, no matter how this issue is fixed, we will definitely upgrade to react-native-web 0.19.x in the future.
And before the upgrade, onModalHide is called in the first task. (the current implementation)
After the upgrade, onModalHide will be called in the second task due to [automatic batching](https://legacy.reactjs.org/blog/2022/03/29/react-v18.html#new-feature-automatic-batching).

@ntdiary Is this mean that this issue will resolve by upgrading RNW?

**Comment by cristipaval** [comment]
Hi @s77rt ! I see in [this P/S](https://expensify.slack.com/archives/C01GTK53T8Q/p1680888088447889?thread_ts=1680398723.148189&cid=C01GTK53T8Q) that you might deal with the composer focus. Do you think that might impact this issue somehow?

**Comment by s77rt** [comment]
@cristipaval Hi :wave: That P/S is slightly related but will not fix the bug. The root cause of this issue is different than the mentioned problem. So let's not hold this and move forward :+1:

**Comment by cristipaval** [comment]
I think we should put this on hold until we upgrade RN. I think we want to go with the best solution, no matter how long it takes. What do you all think?

**Comment by mollfpr** [comment]
I agree 👍

**Comment by cristipaval** [comment]
> Agree, it seems we already have a tracking issue #16660

Great! Thank you, guys! I will put this one on hold and I'm going to also assign @ntdiary to it for now. If the upgrade solves the issue, I think @ntdiary should get paid for the long investigation he has been doing.

**Comment by MelvinBot** [comment]
📣 @ntdiary You have been assigned to this job by @cristipaval!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by davidcardoza** [comment]
I am going to remove myself from the issue since I left the BugZero team.

**Comment by cristipaval** [comment]
Still on hold.

**Comment by cristipaval** [comment]
Still on hold

**Comment by cristipaval** [comment]
Still on hold.

**Comment by cristipaval** [comment]
Still on hold.

**Comment by cristipaval** [comment]
Still on hold.

**Comment by cristipaval** [comment]
> Agree, it seems we already have a tracking issue #16660

Still on hold.

**Comment by cristipaval** [comment]
Nothing changed.

**Comment by cristipaval** [comment]
Same status.

**Comment by cristipaval** [comment]
Making this one monthly as it seems we're moving slowly with the issue that this one is held on.

**Comment by cristipaval** [comment]
Still on hold.

**Comment by cristipaval** [comment]
same status

**Comment by cristipaval** [comment]
> Agree, it seems we already have a tracking issue #16660

Still held on https://github.com/Expensify/App/issues/16660

**Comment by cristipaval** [comment]
This is off hold! @ntdiary @mollfpr could you please check this one again?

**Comment by mollfpr** [comment]
I can't reproduce it either.

**Comment by cristipaval** [comment]
Ok, according to [this comment](https://github.com/Expensify/App/issues/14848#issuecomment-1523278211), I am proposing to pay @mollfpr and @ntdiary for the long investigation that was done several months ago. I'll assign a bugzero member.

**Comment by cristipaval** [comment]
@sonialiap, could you please help me with [this](https://github.com/Expensify/App/issues/14848#issuecomment-1850432608)?

**Comment by sonialiap** [comment]
@cristipaval we typically pay around 25% of the issue price in these cases. Split between the two, does $250 each sound reasonable to you?

**Comment by cristipaval** [comment]
> @cristipaval we typically pay around 25% of the issue price in these cases. Split between the two, does $250 each sound reasonable to you?

yes, thank you @sonialiap! 🙏

**Comment by sonialiap** [comment]
@mollfpr $250 - offer sent 
@ntdiary $250 - offer sent

**Comment by cristipaval** [comment]
What's the payment status here? Can we close the issue?

**Comment by mollfpr** [comment]
@sonialiap I'll do manual request on NewDot, thank you!

**Comment by sonialiap** [comment]
@mollfpr sounds good, withdrawing the Upwork offer 
@ntdiary payment completed 💰

**Comment by JmillsExpensify** [comment]
$250 payment approved for @mollfpr based on [this comment](https://github.com/Expensify/App/issues/14848#issuecomment-1858036542).
